text_generation/causal_lm/cpp/continuous_batching/library/src/sampler.hpp

```diff
@@ -357,35 +366,49 @@ SamplerOutput Sampler::sample(std::vector<SequenceGroup::Ptr> & sequence_groups,
+                std::vector<int64_t> sampled_token_ids;
                 std::vector<Sequence::Ptr> running_sequences = sequence_group->get_running_sequences();
-                OPENVINO_ASSERT(running_sequences.size() == 1);
-
-                int64_t sampled_token_id;
                 if (sampling_params.is_greedy_sampling()) {
-                    sampled_token_id = _greedy_sample(sequence_group_logits);
-                }
-                else {  // .is_multinomial()
-                    sampled_token_id = _multinomial_sample(sequence_group_logits, sampling_params.temperature, sampling_params.top_p, sampling_params.top_k);
+                    OPENVINO_ASSERT(running_sequences.size() == 1);
+                    sampled_token_ids.push_back(_greedy_sample(sequence_group_logits));
+                } else {
+                    // is_multinomial()
+                    auto n = sequence_group->get_num_processed_tokens() > 0 ? 1 : sampling_params.num_return_sequences;
+                    sampled_token_ids = _multinomial_sample(sequence_group_logits, sampling_params.temperature, sampling_params.top_p, sampling_params.top_k, n);
+                    
+                    if (n > 1) {
+                        const auto sequence_to_fork = running_sequences[0];
+                        std::list<uint64_t> forked_seq_ids;
+                        for (; num_running_sequences < n; ++num_running_sequences) {
+                            const auto forked_sequence = sequence_group->fork_sequence(sequence_to_fork);
+                            forked_seq_ids.push_back(forked_sequence->get_id());
+                            running_sequences.push_back(forked_sequence);
+                        }
+                        sampler_output.m_forked_sequences.insert({running_sequences[0]->get_id(), forked_seq_ids});
+                    }
                 }
-                // in case of greedy search we always have a single parent sequence to sample from
-                running_sequences[0]->append_token(sampled_token_id, sequence_group_logits.data<const float>()[sampled_token_id]);
-
-                if (sampling_params.max_new_tokens == running_sequences[0]->get_generated_len() ||
-                    sampled_token_id == sampling_params.eos_token_id && !sampling_params.ignore_eos) {
-                    // stop sequence by max_new_tokens or EOS token
-                    running_sequences[0]->set_status(SequenceStatus::FINISHED);
-                    // drop sequence from scheduler
-                    sampler_output.m_dropped_sequences.push_back(running_sequences[0]->get_id());
+                for (size_t i = 0; i < num_running_sequences; ++i) {
+                    // in case of greedy search we always have a single parent sequence to sample from
+                    auto sampled_token_id = sampled_token_ids[i];
+                    running_sequences[i]->append_token(sampled_token_id, sequence_group_logits.data<const float>()[sampled_token_id]);
```

Comment:
I suppose on next iterations (when child sequences have >= 1 tokens), `sampled_token_id` corresponds to `logits` of this sequence (not to entire sequence group as on first sampling)